File: Project/Simulator/Organisms/Animals/Fox.java
package Simulator.Organisms.Animals;

import Simulator.World;
import Simulator.Tile;
import Utils.RNG;
import java.util.List;
import java.util.ArrayList;

import org.json.JSONObject;

public class Fox extends Animal {

    public Fox(World world) {
        super(3, 7, world, Type.FOX);
    }

    @Override
    public Animal construct() {
        return new Fox(world);
    }

    public Fox(JSONObject json, World world) {
        super(json, world);
    }

    @Override
    public void action() {
        List<Tile> neighbours = new ArrayList<>(tile.getNeighbours()); // Use a new list to avoid
                                                                       // ConcurrentModificationException
        neighbours.removeIf(neighbour -> neighbour == null || !neighbour.isFree()
                || (neighbour.getOrganism() != null && neighbour.getOrganism().getPower() > this.power));

        Tile target;
        if (neighbours.isEmpty()) {
            target = tile.getRandomFreeNeighbour();
        } else {
            target = neighbours.get(RNG.roll(0, neighbours.size() - 1));
        }
        move(target);
    }
}

File: Project/Simulator/Organisms/Animals/Fox.py
from .Animal import Animal
from Simulator.Organisms.Organism import Type

import random


class Fox(Animal):
    def __init__(self, world, json=None):
        super().__init__(3, 7, world, Type.FOX, json)

    def construct(self):
        return Fox(self._world)

    def action(self):
        neighbours = [neighbour for neighbour in self._tile.getNeighbours() if neighbour and neighbour.isFree(
        ) and (not neighbour.getOrganism() or neighbour.getOrganism().getPower() <= self._power)]

        if not neighbours:
            target = self._tile.getRandomFreeNeighbour()
        else:
            target = random.choice(neighbours)

        self.move(target)
